lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,234 @@
+                             , _fields =
+                               entry ^. E.fields & traverse %~
+                                 (\f ->
+                                 FieldMetadata
+                                 { _dateModified = f ^. E.dateModified
+                                 , _private = f ^. E.private
+                                 })
+                               & HS.mapKeys E.getFieldKey
+                             , _tags = map E.getEntryTag $ entry ^. E.tags
+                             }
+        let secret = I.PostSecret
+              { I._cas = Nothing
+              , I._pdata =
+                    HS.insert "#$coffer" (TL.toStrict . TL.decodeUtf8 $ A.encode cofferSpecials)
+                  . HS.map (^. E.value)
+                  . HS.mapKeys E.getFieldKey
+                  $ entry ^. E.fields
+              }
+
+        response <- embedCatch (entry ^. E.path) (postSecret env (entry ^. E.path) secret)
+
+        case response ^. I.ddata . at ("version" :: T.Text) of
+          Just (A.Number i) -> maybe (throw MarshallingFailed) pure (S.toBoundedInteger i)
+          _ -> throw MarshallingFailed
+      ReadSecret path version ->
+        runMaybe $ embedCatchMaybe path (readSecret env path version)
+        >>= \(I.KvResponse _ _ _ _ (I.ReadSecret _data _ _ _ _ _) _ _ _) -> do
+          cofferSpecials :: CofferSpecials <-
+            maybeThrow (_data ^.at "#$coffer" >>= A.decodeStrict' . T.encodeUtf8)
+          let secrets = HS.toList $ foldr HS.delete _data ["#$coffer"]
+          let keyToField key = do
+               _modTime <- cofferSpecials ^. fields.at key <&> (^. dateModified)
+               _private <- cofferSpecials ^. fields.at key <&> (^. private)
+               _value <- _data ^.at key
+               _key <- E.newFieldKey key
+
+               Just (_key
+                    , E.newField _modTime _value
+                      & E.private .~ _private
+                    )
+
+          fields <- maybeThrow $
+            secrets & traverse %~ (keyToField . fst) & sequence <&> HS.fromList
+          _tags <- maybeThrow $ cofferSpecials ^. tags & mapM E.newEntryTag
+
+          pure $ E.newEntry path (cofferSpecials ^. globalDateModified)
+            & E.masterField .~ (cofferSpecials ^. masterKey >>= E.newFieldKey)
+            & E.fields .~ fields
+            & E.tags .~ _tags
+      ListSecrets path ->
+        runMaybe (embedCatchMaybe path (listSecrets env path) <&> (^. I.ddata) <&> \(I.ListSecrets list) -> list)
+      DeleteSecret path ->
+        embedCatch path (deleteSecret env path) <&> const ()
+  where postSecret env = (I.routes env ^. I.postSecret) mount token
+        readSecret env = (I.routes env ^. I.readSecret) mount token
+        listSecrets env = (I.routes env ^. I.listSecrets) mount token
+        updateMetadata env = (I.routes env ^. I.updateMetadata) mount token
+        deleteSecret env = (I.routes env ^. I.deleteSecret) mount token
```

Comment:
These four should be one
```hs
perform env act = (I.routes env ^. act) mount token
```

Reply:
```haskell
lib/Vault/Kv.hs:129:16: error:
    • Couldn't match expected type ‘[T.Text]
                                    -> IO (I.KvResponse I.ListSecrets)’
                  with actual type ‘t0’
        ‘t0’ is untouchable
          inside the constraints: x ~ [T.Text]
          bound by a pattern with constructor:
                     ListSecrets :: forall k (m :: k). [T.Text] -> Vault m [T.Text],
                   in a case alternative
          at lib/Vault/Kv.hs:128:7-22
    • In the first argument of ‘embed’, namely
        ‘((perform I.listSecrets) path)’
      In the first argument of ‘(<&>)’, namely
        ‘embed ((perform I.listSecrets) path)’
      In the first argument of ‘(<&>)’, namely
        ‘embed ((perform I.listSecrets) path) <&> (^. I.ddata)’
    • Relevant bindings include
        perform :: Getting
                     (T.Text -> I.VaultToken -> t0)
                     (Vault.Kv.Internal.Routes (Servant.Client.Generic.AsClientT IO))
                     (T.Text -> I.VaultToken -> t0)
                   -> t0
          (bound at lib/Vault/Kv.hs:70:7)
    |
129 |         embed ((perform I.listSecrets) path) <&> (^. I.ddata) <&> \(I.ListSecrets list) -> list
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build coffer-0.1.0.0 (which is required by exe:coffer from
coffer-0.1.0.0).
```

I'm getting that error, but I don't get at all why I'm getting it

Reply:
> `untouchable`

That means `forall`-s are involved and the function just requires an explicit signature.